arangod/Aql/EngineInfoContainerDBServerServerBased.cpp

```diff
@@ -274,117 +396,128 @@ Result EngineInfoContainerDBServerServerBased::buildEngines(
+            // error as we need to abort after.
+            // Otherwise we need to report 
+            Result res{TRI_ERROR_NO_ERROR};
+            for (auto const& tryRes : responses) {
+              auto response = tryRes.get();
+              if (response.fail()) {
+                if (response.isNot(TRI_ERROR_LOCK_TIMEOUT)) {
+                  // Found something we cannot recover from.
+                  // Return and give up
+                  return response;
+                }
+                // track that we have lock_timeout_present
+                res = response;
+              }
+            }
+            // Return what we have, this will be ok() if and only
+            // if none of the requests failed.
+            // If will be LOCK_TIMEOUT if and only if the only error
+            // we see was LOCK_TIMEOUT.
+            return res;
+          });
+  if (fastPathResult.get().fail()) {
+    if (fastPathResult.get().isNot(TRI_ERROR_LOCK_TIMEOUT)) {
+      return fastPathResult.get();
     }
-    QueryId globalId = 0;
-    auto result = parseResponse(response, snippetIds, server, serverDest,
-                                didCreateEngine, globalId);
-    if (!result.ok()) {
-      return result;
+    {
+      // in case of fast path failure, we need to cleanup engines
+      auto requests = cleanupEngines(fastPathResult.get().errorNumber(), _query.vocbase().name(), serverToQueryId);
+      // Wait for all requests to complete.
+      // So we know that all Transactions are aborted.
+      // We do NOT care for the actual result.
+      futures::collectAll(requests).wait();
+      snippetIds.clear();
+    }
+
+    // set back to default lock timeout for slow path fallback
+    _query.setLockTimeout(oldLockTimeout);
+    LOG_TOPIC("f5022", DEBUG, Logger::AQL)
+        << "Potential deadlock detected, using slow path for locking. This "
+           "is expected if exclusive locks are used.";
+
+    trx.state()->coordinatorRerollTransactionId();
+
+    // Make sure we always use the same ordering on servers
+    std::sort(engineInformation.begin(), engineInformation.end(),
+              [](auto const& lhs, auto const& rhs) {
+                // Entry <0> is the Server
+                return TransactionState::ServerIdLessThan(std::get<0>(lhs), std::get<0>(rhs));
+              });
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+    // Make sure we always maintain the correct ordering of servers
+    // here, if we contact them in increasing name, we avoid dead-locks
+    std::string serverBefore = "";
+#endif
+    // fallback routine, use synchronous requests (slowPath)
+    for (auto& [server, buffer, didCreateEngine] : engineInformation) {
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+      // If the serverBefore has a smaller ID we allways contact by increasing
+      // ID here.
+      TRI_ASSERT(TransactionState::ServerIdLessThan(serverBefore, server));
+      serverBefore = server;
+#endif
```

Comment:
Could be simplified a lot by using something like 
```
TRI_ASSERT(std::is_sorted(engineInformation.begin(), engineInformation.end(), [](auto const& lhs, auto const% rhs) {
  return std::get<0>(lhs) < std::get<1>(lhs);
}));
```
Just an idea. It is a cosmetic change and not necessary now.